File: examples/ReadAfterWrite.java
public class ReadAfterWrite implements java.io.Serializable {
  protected Integer _var24;
  protected java.util.ArrayList<Integer > _var38;
  public ReadAfterWrite() {
    clear();
  }

  public ReadAfterWrite(Integer x, java.util.ArrayList<Integer > l) {
    _var24 = x;
    _var38 = l;
  }
  public void clear() {
    _var24 = 0;
    _var38 = new java.util.ArrayList<Integer > ();
  }
  public Integer getx() {
    return _var24;
  }
  public void elems(java.util.function.Consumer<Integer > _callback) {
    for (Integer _x104 : _var38) {
      _callback.accept(_x104);
    }
  }
  public void do_thing(Integer n) {
    java.util.ArrayList<Integer > _var105 = new java.util.ArrayList<Integer > ();
    java.util.HashMap<Integer , Integer > _histogram107 = new java.util.HashMap<Integer , Integer > ();
    for (Integer _x108 : _var38) {
      Integer _count109 = _histogram107.getOrDefault(_x108, 0);
      _count109 = (_count109 + 1);
      _histogram107.put(_x108, _count109);

    }
    {
      Integer _count109 = _histogram107.getOrDefault((_var24 + n), 0);
      _count109 = (_count109 + 1);
      _histogram107.put((_var24 + n), _count109);

    }
    for (Integer _var106 : _var38) {
      if ((_histogram107.getOrDefault(_var106, 0) > 0)) {
        Integer _count110 = _histogram107.getOrDefault(_var106, 0);
        _count110 = (_count110 - 1);
        _histogram107.put(_var106, _count110);

      } else {
        _var105.add(_var106);
      }
    }
    java.util.ArrayList<Integer > _var101 = _var105;
    java.util.ArrayList<Integer > _var111 = new java.util.ArrayList<Integer > ();
    java.util.HashMap<Integer , Integer > _histogram113 = new java.util.HashMap<Integer , Integer > ();
    for (Integer _x114 : _var38) {
      Integer _count115 = _histogram113.getOrDefault(_x114, 0);
      _count115 = (_count115 + 1);
      _histogram113.put(_x114, _count115);

    }
    for (Integer _var112 : _var38) {
      if ((_histogram113.getOrDefault(_var112, 0) > 0)) {
        Integer _count116 = _histogram113.getOrDefault(_var112, 0);
        _count116 = (_count116 - 1);
        _histogram113.put(_var112, _count116);

      } else {
        _var111.add(_var112);
      }
    }
    {
      Integer _var112 = (_var24 + n);
      if ((_histogram113.getOrDefault(_var112, 0) > 0)) {
        Integer _count116 = _histogram113.getOrDefault(_var112, 0);
        _count116 = (_count116 - 1);
        _histogram113.put(_var112, _count116);

      } else {
        _var111.add(_var112);
      }
    }
    java.util.ArrayList<Integer > _var102 = _var111;
    Integer _var103 = (_var24 + n);
    for (Integer _var57 : _var101) {
      _var38.remove(_var57);
    }
    for (Integer _var57 : _var102) {
      _var38.add(_var57);
    }
    _var24 = _var103;
  }
}

File: examples/read-after-write.cpp
#pragma once
#include <algorithm>
#include <functional>
#include <vector>
#include <unordered_set>
#include <string>
#include <unordered_map>

class ReadAfterWrite {
public:
protected:
  int _var24;
  std::vector< int  > _var38;
public:
  inline ReadAfterWrite() {
    _var24 = 0;
    _var38 = (std::vector< int  > ());
  }
  explicit inline ReadAfterWrite(int x, std::vector< int  > l) {
    _var24 = x;
    _var38 = l;
  }
  ReadAfterWrite(const ReadAfterWrite& other) = delete;
  inline int  getx() {
    return _var24;
  }
  template <class F>
  inline void elems(const F& _callback) {
    for (int _x117 : _var38) {
      _callback(_x117);
    }
  }
  inline void do_thing (int n) {
    std::vector< int  > _var118 = (std::vector< int  > ());
    std::unordered_map< int , int , std::hash<int > > _histogram120 = (std::unordered_map< int , int , std::hash<int > > ());
    for (int _x121 : _var38) {
      std::unordered_map< int , int , std::hash<int > >::iterator _map_iterator130 = _histogram120.find(_x121);
      if ((_map_iterator130 == _histogram120.end())) {
        _map_iterator130 = (_histogram120.emplace(_x121, 0).first);
      }
      int &_count122 = _map_iterator130->second;

      _count122 = (_count122 + 1);
    }
    {
      std::unordered_map< int , int , std::hash<int > >::iterator _map_iterator131 = _histogram120.find((_var24 + n));
      if ((_map_iterator131 == _histogram120.end())) {
        _map_iterator131 = (_histogram120.emplace((_var24 + n), 0).first);
      }
      int &_count122 = _map_iterator131->second;

      _count122 = (_count122 + 1);
    }
    for (int _var119 : _var38) {
      std::unordered_map< int , int , std::hash<int > >::const_iterator _map_iterator132 = (_histogram120.find(_var119));
      int _v133;
      if ((_map_iterator132 == _histogram120.end())) {
        _v133 = 0;
      } else {
        _v133 = (_map_iterator132->second);
      }
      if ((_v133 > 0)) {
        std::unordered_map< int , int , std::hash<int > >::iterator _map_iterator134 = _histogram120.find(_var119);
        if ((_map_iterator134 == _histogram120.end())) {
          _map_iterator134 = (_histogram120.emplace(_var119, 0).first);
        }
        int &_count123 = _map_iterator134->second;

        _count123 = (_count123 - 1);
      } else {
        _var118.push_back(_var119);
      }
    }
    std::vector< int  > _var101 = std::move(_var118);
    std::vector< int  > _var124 = (std::vector< int  > ());
    std::unordered_map< int , int , std::hash<int > > _histogram126 = (std::unordered_map< int , int , std::hash<int > > ());
    for (int _x127 : _var38) {
      std::unordered_map< int , int , std::hash<int > >::iterator _map_iterator135 = _histogram126.find(_x127);
      if ((_map_iterator135 == _histogram126.end())) {
        _map_iterator135 = (_histogram126.emplace(_x127, 0).first);
      }
      int &_count128 = _map_iterator135->second;

      _count128 = (_count128 + 1);
    }
    for (int _var125 : _var38) {
      std::unordered_map< int , int , std::hash<int > >::const_iterator _map_iterator136 = (_histogram126.find(_var125));
      int _v137;
      if ((_map_iterator136 == _histogram126.end())) {
        _v137 = 0;
      } else {
        _v137 = (_map_iterator136->second);
      }
      if ((_v137 > 0)) {
        std::unordered_map< int , int , std::hash<int > >::iterator _map_iterator138 = _histogram126.find(_var125);
        if ((_map_iterator138 == _histogram126.end())) {
          _map_iterator138 = (_histogram126.emplace(_var125, 0).first);
        }
        int &_count129 = _map_iterator138->second;

        _count129 = (_count129 - 1);
      } else {
        _var124.push_back(_var125);
      }
    }
    {
      int _var125 = (_var24 + n);
      std::unordered_map< int , int , std::hash<int > >::const_iterator _map_iterator139 = (_histogram126.find(_var125));
      int _v140;
      if ((_map_iterator139 == _histogram126.end())) {
        _v140 = 0;
      } else {
        _v140 = (_map_iterator139->second);
      }
      if ((_v140 > 0)) {
        std::unordered_map< int , int , std::hash<int > >::iterator _map_iterator141 = _histogram126.find(_var125);
        if ((_map_iterator141 == _histogram126.end())) {
          _map_iterator141 = (_histogram126.emplace(_var125, 0).first);
        }
        int &_count129 = _map_iterator141->second;

        _count129 = (_count129 - 1);
      } else {
        _var124.push_back(_var125);
      }
    }
    std::vector< int  > _var102 = std::move(_var124);
    int _var103 = (_var24 + n);
    for (int _var57 : _var101) {
      auto _it142(::std::find(_var38.begin(), _var38.end(), _var57));
      if (_it142 != _var38.end()) { _var38.erase(_it142); }
    }
    for (int _var57 : _var102) {
      _var38.push_back(_var57);
    }
    _var24 = _var103;
  }
};
